### PR TITLE
Liveblog blocks on fronts - AB test

### PR DESCRIPTION
**Pull request by stephanfowler**
Ajax in the latest blocks for blogs. On first view of a liveblog on a front, slide animate the top block. Store the date of the last seen top block in sessionStorage. On subsequent views, slide in any newly published blocks. Numbers of blocks show depends on card size. 

NB: because it's an AB test, there's a flash as the trailtext (if the card includes it) is swapped out for the liveblog blogs.  

With @johnduffell 

liveblog-animation

### CODE REVIEW

**Comment by robertberry** [comment]
... Why have you written another live blog body parser? I told you a library exists for this!

https://github.com/guardian/flexible-content-body-parser

File: static/src/javascripts/projects/facia/modules/ui/live-blog-updates.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
+define([
```
Comment (robertberry) [inline static/src/javascripts/projects/facia/modules/ui/live-blog-updates.js]: Why not ES6? :cry:

File: static/src/javascripts/projects/facia/modules/ui/live-blog-updates.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    var forgetAfterHours = 24,
+        numDisplayedBlocks = 4,
+        blockHeightPx = 44,
+
+        refreshSecs = 60,
+        refreshDecay = 2,
+        refreshMaxTimes = 3,
+
+        selector = '.js-liveblog-blocks',
+        blocksClassname = 'fc-item__liveblog-blocks',
```
Comment (robertberry) [inline static/src/javascripts/projects/facia/modules/ui/live-blog-updates.js]: you should probably capitalise 'name' too.

File: static/src/javascripts/projects/facia/modules/ui/live-blog-updates.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+        if (refreshMaxTimes) {
+            refreshMaxTimes -= 1;
+            setTimeout(updateBlocks, refreshSecs * 1000);
+            refreshSecs = refreshSecs * refreshDecay;
+        }
+    }
+
+    function start() {
+        fastdom.read(function () {
```
Comment (robertberry) [inline static/src/javascripts/projects/facia/modules/ui/live-blog-updates.js]: Don't think you need this `read` here actually. Just looking stuff up in the DOM doesn't have to be done in a `read` cycle. What does have to be done is anything that forces layout to be computed (such as looking at the height of an element).

**Comment by gklopper** [comment]
Is there a mobile version or are we just testing desktop? (I don't mind, just asking)

File: article/app/controllers/ArticleController.scala [hunk 4]
```diff
@@ -24,44 +28,84 @@ case class LiveBlogPage(article: LiveBlog, related: RelatedContent) extends Arti
...
 
-  private def isSupported(c: ApiContent) = c.isArticle || c.isLiveBlog || c.isSudoku
-  override def canRender(i: ItemResponse): Boolean = i.content.exists(isSupported)
+  implicit val blockWrites = (
+    (__ \ "id").write[String] ~
+      (__ \ "title").write[Option[String]] ~
+      (__ \ "publishedDateTime").write[DateTime] ~
+      (__ \ "lastUpdatedDateTime").write[Option[DateTime]] ~
+      (__ \ "body").write[String]
+    )(unlift(TextBlock.unapply))
```
Comment (robertberry) [inline article/app/controllers/ArticleController.scala]:
```scala
object TextBlock {
    implicit val jsonWrites = Json.writes[TextBlock]
}
```

**Comment by stephanfowler** [comment]
Mobile version:

liveblog-mobile

File: static/src/javascripts/projects/facia/modules/ui/live-blog-updates.js (hunk 3 above)
Comment (sndrs) [inline static/src/javascripts/projects/facia/modules/ui/live-blog-updates.js]: :+1:

File: static/src/stylesheets/module/facia/item-layouts/_fc-item--full-media-75.scss [hunk 5]
```diff
@@ -101,4 +101,8 @@ Full item with 75% width media.
     .fc-item__avatar__media {
         right: -5%;
     }
+
+    .fc-item__liveblog-blocks {
+        height: 170px;
```
Comment (sndrs) [inline static/src/stylesheets/module/facia/item-layouts/_fc-item--full-media-75.scss]: what is 170? if it's lines of text, you can use the `get-line-height` sass function to get the height of multiples of particular font style's line-height.

File: static/src/stylesheets/module/facia/item-layouts/_fc-item--list.scss [hunk 6]
```diff
@@ -37,4 +37,8 @@ x x x x x x x x x x x x x x x x x x x x x x x x
         margin: 0;
         @include fs-header(1);
     }
+
+    .fc-item__liveblog-blocks {
+        height: 38px;
```
Comment (sndrs) [inline static/src/stylesheets/module/facia/item-layouts/_fc-item--list.scss]: as with https://github.com/guardian/frontend/pull/8820/files#r28137009

File: static/src/stylesheets/module/facia/item-tones/_tone-live.scss [hunk 7]
```diff
@@ -80,3 +80,53 @@
 .tone-live--most-popular .fc-item__kicker {
     color: colour(live-accent);
 }
+
+.fc-item__liveblog-blocks {
+    height: 82px;
```
Comment (sndrs) [inline static/src/stylesheets/module/facia/item-tones/_tone-live.scss]: as with https://github.com/guardian/frontend/pull/8820/files#r28137009

File: static/src/stylesheets/module/facia/item-tones/_tone-live.scss [hunk 8]
```diff
@@ -80,3 +80,53 @@
...
+    overflow: hidden;
+}
+
+.fc-item__liveblog-blocks__inner {
+    transition: transform .5s ease-in-out;
```
Comment (sndrs) [inline static/src/stylesheets/module/facia/item-tones/_tone-live.scss]: should use the mixin

File: static/src/stylesheets/module/facia/item-tones/_tone-live.scss [hunk 9]
```diff
@@ -80,3 +80,53 @@
...
+
+.fc-item__liveblog-block {
+    display: block;
+    position: relative;
+    @include fs-headline(1);
+    color: #ffffff;
+    overflow: hidden;
+    height: $gs-baseline * 3 + 2;
+    margin-bottom: $gs-baseline / 2;
+    transition: height .8s linear;
```
Comment (sndrs) [inline static/src/stylesheets/module/facia/item-tones/_tone-live.scss]: mixin

File: static/src/stylesheets/module/facia/item-tones/_tone-live.scss [hunk 10]
```diff
@@ -80,3 +80,53 @@
 .tone-live--most-popular .fc-item__kicker {
     color: colour(live-accent);
 }
+
+.fc-item__liveblog-blocks {
```
Comment (sndrs) [inline static/src/stylesheets/module/facia/item-tones/_tone-live.scss]: all these styles are global, not behind any `live` tone restriction – should they be?

File: static/src/stylesheets/module/facia/item-tones/_tone-live.scss [hunk 11]
```diff
@@ -80,3 +80,53 @@
...
+.fc-item__liveblog-block__more {
+    position: absolute;
+    bottom: $gs-baseline / 6;
+    right: 0;
+    width: $gs-gutter;
+    padding-left: $gs-gutter / 10;
+    background-color: colour(live-default);
+}
+
+.u-faux-block-link--hover .fc-item__liveblog-block__more {
```
Comment (sndrs) [inline static/src/stylesheets/module/facia/item-tones/_tone-live.scss]: shoudl probably pin this class to `faux` bit on line 11 (`&`, `& .fc-item__liveblog-block__more`)

File: static/src/stylesheets/module/facia/item-tones/_tone-live.scss [hunk 12]
```diff
@@ -80,3 +80,53 @@
...
+    background-color: darken(colour(live-default), 4%);
+}
+
+.fc-item__liveblog-block__time {
+    font-weight: bold;
+}
+
+.fc-item__liveblog-block--new .fc-item__liveblog-block__time:before {
+    content: 'Updated';
+    margin-right: $gs-baseline / 3;
```
Comment (sndrs) [inline static/src/stylesheets/module/facia/item-tones/_tone-live.scss]: use `gs-gutter` for horizontal spacing

**Comment by stephanfowler** [comment]
@sndrs thanks for the comments - I fixed all except the magic number heights. Can't seem to meaningfully express them out of line heights and baselines. The design is changing anyway, so I'm going to keep as is (also running out of testing time before the weekend...)

**Comment by robertberry** [comment]
:+1:

**Pull request merged**